lib/testing/matrix.js

```diff
@@ -0,0 +1,209 @@
+    *
+    * @returns {TestMatrix} new testing Matrix
+    */
+    constructor(params) {
+        /**
+        * Create deep copy for prevent the user to modify the object outside
+        * the class
+        **/
+        this.params = Object.assign({}, params);
+        this.paramsToString = JSON.stringify(this.params);
+        this.callback = null;
+        this.elementsToSpecialize = [];
+        this.description = '';
+        this.listOfExceptions = [];
+    }
+
+    /**
+    * Callback used for prevent mocha that the test is over
+    * Must be call in the end of every test
+    * @callback Done
+    *
+    * @return {undefined}
+    */
+    /**
+    * Callback used for a generated matrix
+    * Used by @generate and @if method
+    *
+    * @callback GenerateMatrixCallback
+    * @param {TestMatrix} matrix - The following generate matrix.
+    * @param {Done} done - Callback for prevent mocha that we finish
+    * the current test. This callback will be empty if not all elements are
+    * specialize
+    *
+    * @return {undefined}
+    */
+
+    /**
+    * Generate testing matrix, call execute method to run callback
+    *
+    * @param {string[]} elementsToSpecialize - key element we want to
+    *                    specialize on the matrix
+    * @param {GenerateMatrixCallback} callback - callback function
+    *                    who want to be call on execute method
+    *                    for the current generate matrix
+    * @param {string} description - description of the test
+    *
+    * @returns {TestMatrix} Return current instance of Matrix object
+    */
+    generate(elementsToSpecialize, callback, description) {
+        this.elementsToSpecialize = elementsToSpecialize;
+        this.callback = callback;
+        this.description = typeof description === 'undefined'
+        ? ''
+        : description;
+        return this;
+    }
+
+    /**
+    * Method to prevent an exception during the tests.
+    *
+    * @param {Object.<string, string[]>} exception - The keys of the hash table
+    is the element we want to do an exception, the value
+    *                      is the value we want to do an exception for the
+    *                      specific key.
+    *                      All key of the object must be validate with at least
+    *                      one value for execute an exception.
+    *                      ex: for a hashmap with 'auth' as key and 'v4' as
+    *                      value, if we found a matrix with this following
+    *                      paramater, it will call the special callback
+    * @param {GenerateMatrixCallback} callbackException - Callback who the
+    *                       execute method will call when
+    *                       this exception is found.
+    * @param {string} description - the test's description
+    *
+    * @returns {TestMatrix} Return current instance of Matrix object
+    */
+    if(exception, callbackException, description) {
```

Comment:
The callback isn't the last argument because the description is optionnal. Which is useful for intermediate generated matrix who haven't have any description because it's not a really a test as himself, it's only the final generated matrix who is a test.